### PR TITLE
slim.learning.train and slim.evaluation.evaluation to handle OutOfRange gracefully

### PR DESCRIPTION
**Pull request by alsrgv**
While doing this fix, I noticed that some of the metric operations don't handle OutOfRange gracefully also.  

For example, streaming mean computation would update total value and count concurrently, possibly updating count before failing to update total value due to OutOfRange. 

I have added fixes for mean metrics, but I am not sure what other places might be handling OutOfRange incorrectly.

### CODE REVIEW

**Comment by sguada** [comment]
Please update and resolve the conflicts. 

I think the functionality is welcome.

**Comment by alsrgv** [comment]
@sguada thanks for feedback.  PR is updated.

**Comment by drpngx** [comment]
@alsgrv sorry this fell through the cracks. Could you resolve changes and push again?

**Comment by alsrgv** [comment]
@drpngx merged.

**Comment by drpngx** [comment]
Looks like there's another conflict, could you push again?

**Comment by alsrgv** [comment]
@drpngx, sure - resolved.

File: tensorflow/contrib/slim/python/slim/evaluation_test.py [hunk 1]
```diff
@@ -157,6 +157,33 @@ def testMonitorCheckpointsLoopTimeout(self):
...
+  def testWithEpochLimit(self):
+    predictions_limited = tf.train.limit_epochs(self._predictions, num_epochs=1)
+    labels_limited = tf.train.limit_epochs(self._labels, num_epochs=1)
+
+    value_op, update_op = slim.metrics.streaming_accuracy(
+        predictions_limited, labels_limited)
+
+    init_op = tf.group(tf.global_variables_initializer(),
+                       tf.local_variables_initializer())
+    # Create Checkpoint and log directories
```
Comment (drpngx) [inline tensorflow/contrib/slim/python/slim/evaluation_test.py]: Punctuation please. Use "checkpoint" for consistency (not "Checkpoint").

File: tensorflow/contrib/slim/python/slim/evaluation_test.py [hunk 2]
```diff
@@ -157,6 +157,33 @@ def testMonitorCheckpointsLoopTimeout(self):
...
+    chkpt_dir = os.path.join(self.get_temp_dir(), 'tmp_logs/')
+    gfile.MakeDirs(chkpt_dir)
+    logdir = os.path.join(self.get_temp_dir(), 'tmp_logs2/')
+    gfile.MakeDirs(logdir)
+
+    # Save initialized variables to checkpoint directory
```
Comment (drpngx) [inline tensorflow/contrib/slim/python/slim/evaluation_test.py]: Punctuation.

File: tensorflow/contrib/slim/python/slim/learning.py [hunk 3]
```diff
@@ -788,12 +788,15 @@ def train(train_op,
...
-              break
+          try:
+            while not sv.should_stop():
+              total_loss, should_stop = train_step_fn(sess, train_op, global_step,
+                                                      train_step_kwargs)
+              if should_stop:
+                logging.info('Stopping Training.')
+                break
+          except errors.OutOfRangeError:
+            logging.info('Caught OutOfRangeError. Stopping Training.')
```
Comment (drpngx) [inline tensorflow/contrib/slim/python/slim/learning.py]: It's probably a logging.error, although I'm not sure why you would want to catch that error and recover from it.

File: tensorflow/python/ops/metrics_impl.py [hunk 4]
```diff
@@ -263,8 +263,9 @@ def mean(values, weights=None, metrics_collections=None,
       values = math_ops.multiply(values, weights)
       num_values = math_ops.reduce_sum(weights)
 
-    update_total_op = state_ops.assign_add(total, math_ops.reduce_sum(values))
-    update_count_op = state_ops.assign_add(count, num_values)
+    with ops.control_dependencies([values]):
+      update_total_op = state_ops.assign_add(total, math_ops.reduce_sum(values))
```
Comment (drpngx) [inline tensorflow/python/ops/metrics_impl.py]: So technically, this doesn't ened to be inside the `control_dependencies`, right?

File: tensorflow/python/ops/metrics_impl.py [hunk 5]
```diff
@@ -263,8 +263,9 @@ def mean(values, weights=None, metrics_collections=None,
...
+      update_count_op = state_ops.assign_add(count, num_values)
```
Comment (drpngx) [inline tensorflow/python/ops/metrics_impl.py]: So this is for the else branch, above right? I think it's probably better placed there, because that's the one that introduces the race condition?

File: tensorflow/python/ops/metrics_impl.py [hunk 6]
```diff
@@ -978,8 +979,9 @@ def mean_tensor(values, weights=None, metrics_collections=None,
       values = math_ops.multiply(values, weights)
       num_values = math_ops.multiply(num_values, weights)
 
-    update_total_op = state_ops.assign_add(total, values)
-    update_count_op = state_ops.assign_add(count, num_values)
+    with ops.control_dependencies([values]):
```
Comment (drpngx) [inline tensorflow/python/ops/metrics_impl.py]: Same as above.

**Review by drpngx — CHANGES_REQUESTED** [review]
Some additional questions.

File: tensorflow/contrib/slim/python/slim/evaluation_test.py (hunk 2 above)
Comment (alsrgv) [inline tensorflow/contrib/slim/python/slim/evaluation_test.py]: Fixed

File: tensorflow/contrib/slim/python/slim/evaluation_test.py (hunk 1 above)
Comment (alsrgv) [inline tensorflow/contrib/slim/python/slim/evaluation_test.py]: Fixed

File: tensorflow/contrib/slim/python/slim/learning.py (hunk 3 above)
Comment (alsrgv) [inline tensorflow/contrib/slim/python/slim/learning.py]: OutOfRangeError is thrown when epoch limit per tf.train.limit_epochs is reached.  In my opinion it should be condition to normally finish training.

Comment (drpngx) [inline tensorflow/contrib/slim/python/slim/learning.py]: Ack. Sigh. Thanks for the clarification. Perhaps a comment might help.

Comment (alsrgv) [inline tensorflow/contrib/slim/python/slim/learning.py]: Good point, added comment.

File: tensorflow/python/ops/metrics_impl.py (hunk 5 above)
Comment (alsrgv) [inline tensorflow/python/ops/metrics_impl.py]: Is this new change along the lines of what you're thinking?

**Review by drpngx — APPROVED** [review]

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by alsrgv** [comment]
Looks like evaluation test failed, let me see what's going on.

**Comment by alsrgv** [comment]
Jenkins, test this please.

**Comment by alsrgv** [comment]
@drpngx, I had to put back this guard for both branches:

```
with ops.control_dependencies([values]):
  update_count_op = state_ops.assign_add(count, num_values)
```

This is because this operation doesn't seem to force evaluation of values:

```
math_ops.to_float(array_ops.size(values))
```

Thus, count get incremented while sum is unchanged due to OutOfRangeError.

**Comment by drpngx** [comment]
Oh, got it. That's a bit of a gotcha.

Jenkins, test this please.

**Comment by alsrgv** [comment]
Looks like it failed due to estimator_test timeout.  I checked another run, and it's pretty slow in general:
```
//tensorflow/contrib/learn:estimator_test                                PASSED in 243.6s
```

On my machine is takes ~70 seconds.  Could this be transient failure?

**Comment by drpngx** [comment]
Yeah, it seems to happen in other builds.
